client.go

```diff
@@ -87,11 +240,252 @@ func (c *Client) SendSTUNRequest(serverIP net.IP, serverPort int) (net.Addr, err
 	}, nil
 }
 
-func sendStunRequest(conn net.PacketConn, serverIP net.IP, serverPort int) error {
-	msg, err := stun.Build(stun.TransactionID, stun.BindingRequest)
+// SendBindingRequest sends a new STUN request to the STUN server
+func (c *Client) SendBindingRequest() (net.Addr, error) {
+	if c.stunServ == nil {
+		return nil, fmt.Errorf("STUN server address is not set for the client")
+	}
+	return c.SendBindingRequestTo(c.stunServ)
+}
+
+// Allocate sends a TURN allocation request to the given transport address
+func (c *Client) Allocate() (net.PacketConn, error) {
+	if c.relayedConn != nil {
```

Comment:
I would put a mutex on `Allocate`. A bad user could call Allocate twice, always assume the worst :)

Reply:
Ah, that's true...  I'd need to review entire code for thread-safety.

Reply:
The solution wouldn't be the mutex... After some thoughts, I introduced a simple try-lock, which would return an error immediately when Allocate() was called while another goroutine's Allocate() call is not complete yet. Can you review it again?